File: lib/domain/deletion_undo/dtos/deletion_undo_schedule_payload_failure.rb
# frozen_string_literal: true

module Domain
  module DeletionUndo
    module Dtos
      # ペイロード組み立て時の失敗（例: undo トークン欠落）。ステータスは Presenter がそのまま HTTP に写す。
      class DeletionUndoSchedulePayloadFailure
        attr_reader :reason, :http_status

        def initialize(reason:, http_status: :unprocessable_entity)
          @reason = reason
          @http_status = http_status
        end
      end
    end
  end
end

File: crates/agrr-domain/src/deletion_undo/dtos/deletion_undo_schedule_payload_failure.rs
//! Ruby: `Domain::DeletionUndo::Dtos::DeletionUndoSchedulePayloadFailure`

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DeletionUndoSchedulePayloadFailureReason {
    MissingUndoToken,
}

/// Ruby: `Domain::DeletionUndo::Dtos::DeletionUndoSchedulePayloadFailure`
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DeletionUndoSchedulePayloadFailure {
    pub reason: DeletionUndoSchedulePayloadFailureReason,
    pub http_status: u16,
}

impl DeletionUndoSchedulePayloadFailure {
    pub fn missing_undo_token() -> Self {
        Self {
            reason: DeletionUndoSchedulePayloadFailureReason::MissingUndoToken,
            http_status: 422,
        }
    }
}
